File: framework/ui/src/io/intino/alexandria/ui/displays/components/Stepper.java
package io.intino.alexandria.ui.displays.components;

import io.intino.alexandria.core.Box;
import io.intino.alexandria.schemas.StepperInfo;
import io.intino.alexandria.ui.displays.Component;
import io.intino.alexandria.ui.displays.Display;
import io.intino.alexandria.ui.displays.events.Event;
import io.intino.alexandria.ui.displays.events.Listener;
import io.intino.alexandria.ui.displays.notifiers.StepperNotifier;

import java.util.List;

import static java.util.stream.Collectors.toList;

public class Stepper<DN extends StepperNotifier, B extends Box> extends AbstractStepper<B> {
    private Listener finishListener = null;
    protected int active = 0;

    public Stepper(B box) {
        super(box);
    }

    public Stepper onFinish(Listener listener){
        this.finishListener = listener;
        return this;
    }

    @Override
    public void init() {
        super.init();
        updateActive();
    }

    @Override
    public <D extends Display> D register(D child) {
        D registered = super.register(child);
        updateSteps();
        return registered;
    }

    public void add(Step step) {
        add(step, "steps");
    }

    protected List<Step> steps() {
        return children(Step.class);
    }

    private void updateSteps() {
        int index = 0;
        for (Step step : steps()) {
            step._isActive(index == active);
            step._isCompleted(index < active);
            step.refresh();
            if (index == active) step.show();
            index++;
        }
    }

    public void reset() {
        select(0);
    }

    public void select(Step step) {
        select(posOf(step));
    }

    public void select(int pos) {
        this.active = pos;
        updateActive();
    }

    public void next() {
        if (allowNext()) {
            this.active++;
            updateActive();
            updateSteps();
            checkFinish();
        }
    }

    public void back() {
        if (allowBack()){
            this.active--;
            updateActive();
            updateSteps();
        }
    }

    private boolean allowNext(){
        return visiblePosOfActive() < visibleStepsCount()-1 && steps().get(this.active).allowNext();
    }

    private boolean allowBack() {
        return visiblePosOfActive() > 0 && steps().get(this.active).allowBack();
    }

    private int visiblePosOfActive() {
        if (steps().size() <= 0) return 0;
        int pos = -1;
        List<Boolean> visibleSteps = visibleSteps();
        for (int i=0; i<=this.active; i++) if (visibleSteps.get(i)) pos++;
        return pos;
    }

    private boolean isFinished() {
        return this.active >= visibleStepsCount();
    }

    private void checkFinish() {
        if (isFinished() && finishListener != null) finishListener.accept(new Event(this));
    }

    private void updateActive() {
        notifier.refresh(new StepperInfo().active(active).allowNext(allowNext()).allowBack(allowBack()).visibleList(visibleSteps()));
    }

    private List<Boolean> visibleSteps() {
        return steps().stream().map(Component::isVisible).collect(toList());
    }

    private long visibleStepsCount() {
        return steps().stream().filter(Component::isVisible).count();
    }

    private int posOf(Step step) {
        List<Step> steps = steps();
        for (int i = 0; i< steps.size(); i++) if (steps.get(i) == step) return i;
        return -1;
    }

}

File: framework/ui-elements/src/displays/components/Stepper.js
import React from "react";
import { withStyles } from '@material-ui/core/styles';
import AbstractStepper from "../../../gen/displays/components/AbstractStepper";
import StepperNotifier from "../../../gen/displays/notifiers/StepperNotifier";
import StepperRequester from "../../../gen/displays/requesters/StepperRequester";
import DisplayFactory from 'alexandria-ui-elements/src/displays/DisplayFactory';
import MuiStepper from "@material-ui/core/Stepper";
import MuiButton from "@material-ui/core/Button";
import { withSnackbar } from 'notistack';

const Positions = {
	Top: "top",
	Bottom: "bottom",
	Left: "left",
	Right: "right"
};

const styles = theme => ({
	root:{
		display: "flex"
	},
	row: {
		flexDirection: "row"
	},
	column: {
		flexDirection: "column"
	},
	content: {
		width: "100%"
	}
});

function clsx(position, classes){
	let classNames = [];
	classNames.push(classes.root);
	if (position === Positions.Top || position === Positions.Bottom) classNames.push(classes.column);
	if (position === Positions.Left || position === Positions.Right) classNames.push(classes.row);
	return classNames.join(" ");
}

class Stepper extends AbstractStepper {

	constructor(props) {
		super(props);
		this.notifier = new StepperNotifier(this);
		this.requester = new StepperRequester(this);
		this.state = {
			...this.state,
			active: undefined,
			allowNext: true,
			allowBack: true,
			visibleList: [],
		};
	};

	_isTop(){
		return this.props.position === Positions.Top || this.props.position === Positions.Left;
	};

	render() {
		return (
			<React.Fragment>
				{this._renderContent()}
				{this._renderToolbar()}
			</React.Fragment>
		);
	};

	_renderContent(){
		const { position, classes } = this.props;
		const classNames = clsx(position, classes);
		const isTop =  this._isTop();
		return (
			<div className={classNames} style={{height:"100%"}}>
				{isTop && this._renderStepper()}
				{this._renderActive()}
				{!isTop && this._renderStepper()}
			</div>
		);
	};

	_renderActive(){
		const active = this.state.active;
		const children  = React.Children.toArray(this.props.children);
		const isInRange = this._inRange(active, children.length);
		return (
			<div className={this.props.classes.content} style={{height:"100%",overflow:'auto'}}>
				{ isInRange && React.cloneElement(children[active], { showContent: true })}
			</div>
		);
	};

	_inRange(active, total) {
		return total && active !== undefined && active > -1 && active < total;
	};

	_renderStepper() {
		const { orientation } = this.props;
		const children = this._visibleChildren();
		return (
			<MuiStepper orientation={orientation}>{
				React.Children.map(children,
					(child, i) => React.cloneElement(child, { index: i })
				)
			}</MuiStepper>
		);
	};

	_visibleChildren() {
		const { children } = this.props;
		const visibleList = this.state.visibleList;
		return children.filter((child, i) => visibleList[i] == true);
	};

	_renderToolbar() {
	    if (this.props.style === "NoToolbar") return;
		const count = React.Children.count(this._visibleChildren());
		return (<div>
			<MuiButton onClick={() => this.requester.back()} disabled={this.state.active <= 0 || !this.state.allowBack}>{this.translate("Back")}</MuiButton>
			<MuiButton onClick={() => this.requester.next()} disabled={this.state.active >= count || !this.state.allowNext}>{this.translate("Next")}</MuiButton>
		</div>);
	};

	refresh({active, allowNext, allowBack, visibleList}) {
		this.setState({active,allowNext,allowBack,visibleList});
	};

}

export default withStyles(styles, { withTheme: true })(withSnackbar(Stepper));
DisplayFactory.register("Stepper", withStyles(styles, { withTheme: true })(withSnackbar(Stepper)));